lib/src/command/run_command.dart

```diff
@@ -0,0 +1,121 @@
+        this._downloader = downloader ?? Downloader() {
+    _RunArgResults.addOptions(argParser);
+  }
+
+  @override
+  FutureOr<Null> run() async {
+    final rootDirectory = sonarScannerArgResults.workingDirectory ?? Directory.current.path;
+    final cacheDirectory = sonarScannerArgResults.cacheDirectory ?? _getCacheDirectoryPath();
+    final distributionName = _extractDistributionName(_runArgResults.distributionUrl);
+    final distributionDirectory = Directory("${cacheDirectory}/${distributionName}");
+    if (_shouldDownloadDistribution(distributionDirectory)) {
+      final archive = await _downloadDistributionArchive(_runArgResults.distributionUrl);
+      _extractArchive(archive, cacheDirectory);
+    }
+    final processExitCode = await _startProcess(
+      distributionDirectory.path,
+      argResults.rest,
+      rootDirectory,
+    ).then((process) => process.exitCode);
+
+    if (processExitCode != 0) {
+      exitCode = processExitCode;
+    }
+  }
+
+  String _getCacheDirectoryPath() {
+    final basePath = Platform.environment[Platform.isWindows ? "APPDATA" : "HOME"];
+    return Directory(path.join(basePath, ".sonar-scanner")).path;
+  }
+
+  String _extractDistributionName(String distributionUrl) {
+    return "sonar-scanner-${RegExp(r"cli-(.+).zip").firstMatch(distributionUrl).group(1)}";
+  }
+
+  bool _shouldDownloadDistribution(Directory distributionDirectory) {
+    return _runArgResults.clearCache || !distributionDirectory.existsSync();
+  }
+
+  Future<Archive> _downloadDistributionArchive(String distributionUrl) async {
+    final data = await _downloader.download(distributionUrl);
+    return ZipDecoder().decodeBytes(data);
+  }
+
+  void _extractArchive(Archive archive, String destination) {
+    for (final file in archive) {
+      if (file.isFile) {
+        File("${destination}/${file.name}")
+          ..createSync(recursive: true)
+          ..writeAsBytesSync(file.content as List<int>);
+      } else {
+        Directory("${destination}/${file.name}")..createSync(recursive: true);
+      }
+    }
+  }
+
+  Future<Process> _startProcess(
+    String distributionPath,
+    Iterable<String> arguments,
+    String workingDirectory,
+  ) {
+    final command = Platform.isWindows ? "${distributionPath}/bin/sonar-scanner.bat" : "sh";
+    final platformArguments =
+        Platform.isWindows ? <String>[] : ["${distributionPath}/bin/sonar-scanner"];
+    return _processStarter.start(
+      command,
+      platformArguments.followedBy(arguments).toList(),
+      workingDirectory,
+    );
+  }
+}
+
+class _RunArgResults {
+  final String distributionUrl;
+  final bool clearCache;
+
+  _RunArgResults.fromArgResults(ArgResults results)
+      : this.distributionUrl = fromResults(results, "distribution-url"),
+        this.clearCache = fromResults(results, "clear-cache");
+
+  static void addOptions(ArgParser argParser) {
+    argParser.addOption(
+      "distribution-url",
```

Comment:
What do you think about extract "distribution-url" and "clear-cache" to final private variables?